### PR TITLE
Add bench-cpu

### PR DESCRIPTION
**Pull request by Dr-Electron**
# Description of change

Ported the `bench-cpu` feature from [Hornet](https://github.com/gohornet/hornet/blob/main/pkg/toolset/benchmark_cpu.go)

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Tested locally but somehow I still get really bad results compared to Hornet. Pretty sure I miss something obvious.
Hornet results:
```
./hornet tools bench-cpu
Average CPU speed: 7.20MH/s (16 thread(s), 3.34%. 58s left...)
Average CPU speed: 7.16MH/s (16 thread(s), 6.67%. 56s left...)
Average CPU speed: 7.15MH/s (16 thread(s), 10.04%. 53s left...)
Average CPU speed: 7.10MH/s (16 thread(s), 13.39%. 51s left...)
Average CPU speed: 6.97MH/s (16 thread(s), 16.76%. 49s left...)
Average CPU speed: 6.83MH/s (16 thread(s), 20.17%. 47s left...)
Average CPU speed: 6.76MH/s (16 thread(s), 23.52%. 45s left...)
Average CPU speed: 6.71MH/s (16 thread(s), 26.86%. 43s left...)
Average CPU speed: 6.69MH/s (16 thread(s), 30.20%. 41s left...)
Average CPU speed: 6.64MH/s (16 thread(s), 33.57%. 39s left...)
Average CPU speed: 6.60MH/s (16 thread(s), 36.91%. 37s left...)
Average CPU speed: 6.58MH/s (16 thread(s), 40.25%. 35s left...)
Average CPU speed: 6.55MH/s (16 thread(s), 43.60%. 33s left...)
Average CPU speed: 6.51MH/s (16 thread(s), 46.96%. 31s left...)
Average CPU speed: 6.50MH/s (16 thread(s), 50.33%. 29s left...)
Average CPU speed: 6.48MH/s (16 thread(s), 53.67%. 27s left...)
Average CPU speed: 6.46MH/s (16 thread(s), 57.01%. 25s left...)
Average CPU speed: 6.44MH/s (16 thread(s), 60.39%. 23s left...)
Average CPU speed: 6.43MH/s (16 thread(s), 63.72%. 21s left...)
Average CPU speed: 6.42MH/s (16 thread(s), 67.09%. 19s left...)
Average CPU speed: 6.40MH/s (16 thread(s), 70.44%. 17s left...)
Average CPU speed: 6.39MH/s (16 thread(s), 73.80%. 15s left...)
Average CPU speed: 6.38MH/s (16 thread(s), 77.16%. 13s left...)
Average CPU speed: 6.37MH/s (16 thread(s), 80.53%. 11s left...)
Average CPU speed: 6.36MH/s (16 thread(s), 83.87%. 9s left...)
Average CPU speed: 6.35MH/s (16 thread(s), 87.22%. 7s left...)
Average CPU speed: 6.33MH/s (16 thread(s), 90.59%. 5s left...)
Average CPU speed: 6.33MH/s (16 thread(s), 93.93%. 3s left...)
Average CPU speed: 6.32MH/s (16 thread(s), 97.30%. 1s left...)
Average CPU speed: 6.32MH/s (16 thread(s), took 1m0s)

./hornet tools bench-cpu 1
Average CPU speed: 1.07MH/s (1 thread(s), 3.34%. 58s left...)
Average CPU speed: 1.08MH/s (1 thread(s), 6.68%. 56s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 10.01%. 53s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 13.36%. 51s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 16.69%. 49s left...)
Average CPU speed: 1.06MH/s (1 thread(s), 20.03%. 47s left...)
Average CPU speed: 1.06MH/s (1 thread(s), 23.37%. 45s left...)
Average CPU speed: 1.06MH/s (1 thread(s), 26.71%. 43s left...)
Average CPU speed: 1.06MH/s (1 thread(s), 30.04%. 41s left...)
Average CPU speed: 1.06MH/s (1 thread(s), 33.37%. 39s left...)
Average CPU speed: 1.06MH/s (1 thread(s), 36.71%. 37s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 40.05%. 35s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 43.40%. 33s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 46.73%. 31s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 50.07%. 29s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 53.41%. 27s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 56.74%. 25s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 60.09%. 23s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 63.42%. 21s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 66.76%. 19s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 70.10%. 17s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 73.44%. 15s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 76.78%. 13s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 80.12%. 11s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 83.46%. 9s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 86.79%. 7s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 90.13%. 5s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 93.47%. 3s left...)
Average CPU speed: 1.07MH/s (1 thread(s), 96.80%. 1s left...)
Average CPU speed: 1.07MH/s (1 thread(s), took 1m0s)
```

Bee results:
```
./bee bench-cpu
Benchmarking CPU with 16 threads
Average CPU speed: 0.26MH/s (16 thread(s), 3.33%. 58.01s left...)
Average CPU speed: 0.26MH/s (16 thread(s), 6.68%. 56.00s left...)
Average CPU speed: 0.26MH/s (16 thread(s), 10.01%. 54.00s left...)
Average CPU speed: 0.25MH/s (16 thread(s), 13.35%. 51.99s left...)
Average CPU speed: 0.25MH/s (16 thread(s), 16.69%. 49.99s left...)
Average CPU speed: 0.25MH/s (16 thread(s), 20.03%. 47.99s left...)
Average CPU speed: 0.24MH/s (16 thread(s), 23.36%. 45.99s left...)
Average CPU speed: 0.24MH/s (16 thread(s), 26.69%. 43.99s left...)
Average CPU speed: 0.24MH/s (16 thread(s), 30.03%. 41.98s left...)
Average CPU speed: 0.24MH/s (16 thread(s), 33.37%. 39.98s left...)
Average CPU speed: 0.24MH/s (16 thread(s), 36.70%. 37.98s left...)
Average CPU speed: 0.24MH/s (16 thread(s), 40.04%. 35.98s left...)
Average CPU speed: 0.24MH/s (16 thread(s), 43.38%. 33.98s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 46.71%. 31.97s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 50.05%. 29.97s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 53.38%. 27.97s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 56.72%. 25.97s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 60.06%. 23.97s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 63.40%. 21.96s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 66.74%. 19.96s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 70.08%. 17.95s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 73.42%. 15.95s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 76.75%. 13.95s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 80.09%. 11.95s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 83.43%. 9.94s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 86.78%. 7.94s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 90.11%. 5.93s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 93.45%. 3.93s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 96.79%. 1.93s left...)
Average CPU speed: 0.23MH/s (16 thread(s), 100.00%. 0.00ns left...)
Average CPU speed: 0.23MH/s (16 thread(s), took 60.00s)
./bee bench-cpu 1
Benchmarking CPU with 1 threads
Average CPU speed: 0.04MH/s (1 thread(s), 3.33%. 58.01s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 6.67%. 56.01s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 10.01%. 54.00s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 13.34%. 52.00s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 16.68%. 49.99s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 20.02%. 47.99s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 23.35%. 45.99s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 26.69%. 43.99s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 30.03%. 41.98s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 33.37%. 39.98s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 36.70%. 37.98s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 40.04%. 35.98s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 43.37%. 33.98s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 46.70%. 31.98s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 50.04%. 29.97s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 53.39%. 27.97s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 56.72%. 25.97s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 60.06%. 23.97s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 63.40%. 21.96s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 66.74%. 19.96s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 70.08%. 17.95s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 73.42%. 15.95s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 76.76%. 13.95s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 80.10%. 11.94s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 83.44%. 9.94s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 86.77%. 7.94s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 90.11%. 5.94s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 93.44%. 3.94s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 96.78%. 1.93s left...)
Average CPU speed: 0.04MH/s (1 thread(s), 100.00%. 0.00ns left...)
Average CPU speed: 0.04MH/s (1 thread(s), took 60.00s)
```

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [ ] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-node/src/tools/benchmark_cpu.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    let counter_2 = counter.clone();
+
+    let time_start = std::time::Instant::now();
+
+    let pow_digest: [u8; 32] = rand::random();
+
+    let mut workers = Vec::with_capacity(threads + 2);
+
+    //Stop if the timeout has exceeded
+    let time_thread = thread::spawn(move || {
```
Comment (pvdrz) [inline bee-node/src/tools/benchmark_cpu.rs:54]: maybe these should be `tokio::task::spawn` instead

Comment (Dr-Electron) [reply bee-node/src/tools/benchmark_cpu.rs:54]: Is there a difference in performance? It was the only thing I knew back then. As I said this code is quite old 😆

Comment (pvdrz) [reply bee-node/src/tools/benchmark_cpu.rs:54]: Well... maybe. Tokio's task are closer to goroutines than OS threads. I guess it depends on how much context switching is the OS doing with the threads or something.

**Comment by grtlr** [comment]
I updated the branch and tried to get rid of allocations by making the code closer to what happens in the node. But in my tests this didn't noticeably change the output.

File: bee-node/src/tools/benchmark_cpu.rs [hunk 2]
```diff
@@ -97,14 +97,16 @@ pub fn exec(tool: &BenchmarkCPUTool) {
...
+    let mut pow_digest = TritBuf::<T1B1Buf>::with_capacity(HASH_LENGTH);
     b1t6::encode::<T1B1Buf>(&_pow_digest)
         .iter()
         .for_each(|t| pow_digest.push(t));
 
     let mut nonce = start_nonce;
-    let mut curlp = CurlP::new(CurlPRounds::Rounds81);
-    let mut buffers = TritBuf::<T1B1Buf>::new();
+    let mut curlp = CurlP81::new();
+    let mut buffers = TritBuf::<T1B1Buf>::with_capacity(HASH_LENGTH);
```
Comment (thibault-martinez) [inline bee-node/src/tools/benchmark_cpu.rs]: Is it really `HASH_LENGTH` ? Seems to me that the final size of `buffers` is `BATCH_SIZE * (pow_digest.len() + nonce_trits.len())` ?

Comment (grtlr) [reply bee-node/src/tools/benchmark_cpu.rs]: It looks like that's what we use in `bee-pow`: https://github.com/iotaledger/bee/blob/c08f1bac7170fe5cb650ddc347ac4d483bb9036a/bee-pow/src/score.rs#L29

**Update** Nevermind, you're right, there is something wrong here.

**Comment by grtlr** [comment]
Ok, so I have updated the dependencies to make use of the recent performance improvements of `bee-ternary`. I've also restructured the benchmark to bring it more in line with how the mining works internally. Through that I was able to achieve the following scores:

```
cargo run --release -- bench-cpu
Benchmarking CPU with 8 threads
Average CPU speed: 73.48MH/s (8 thread(s), 3.33%. 58.00s left...)
Average CPU speed: 74.28MH/s (8 thread(s), 6.67%. 56.00s left...)
Average CPU speed: 75.28MH/s (8 thread(s), 10.00%. 54.01s left...)
Average CPU speed: 75.64MH/s (8 thread(s), 13.33%. 52.00s left...)
Average CPU speed: 76.41MH/s (8 thread(s), 16.67%. 50.00s left...)
Average CPU speed: 76.61MH/s (8 thread(s), 20.01%. 48.00s left...)
Average CPU speed: 76.90MH/s (8 thread(s), 23.35%. 45.99s left...)
Average CPU speed: 77.40MH/s (8 thread(s), 26.68%. 43.99s left...)
Average CPU speed: 77.42MH/s (8 thread(s), 30.03%. 41.99s left...)
Average CPU speed: 77.00MH/s (8 thread(s), 33.36%. 39.99s left...)
Average CPU speed: 77.46MH/s (8 thread(s), 36.69%. 37.98s left...)
Average CPU speed: 77.89MH/s (8 thread(s), 40.03%. 35.98s left...)
Average CPU speed: 78.18MH/s (8 thread(s), 43.37%. 33.98s left...)
Average CPU speed: 77.77MH/s (8 thread(s), 46.71%. 31.97s left...)
Average CPU speed: 78.01MH/s (8 thread(s), 50.04%. 29.97s left...)
Average CPU speed: 78.26MH/s (8 thread(s), 53.38%. 27.97s left...)
Average CPU speed: 78.50MH/s (8 thread(s), 56.72%. 25.97s left...)
Average CPU speed: 78.71MH/s (8 thread(s), 60.05%. 23.97s left...)
Average CPU speed: 78.90MH/s (8 thread(s), 63.39%. 21.97s left...)
Average CPU speed: 79.08MH/s (8 thread(s), 66.72%. 19.97s left...)
Average CPU speed: 79.21MH/s (8 thread(s), 70.06%. 17.97s left...)
Average CPU speed: 79.23MH/s (8 thread(s), 73.40%. 15.96s left...)
Average CPU speed: 78.50MH/s (8 thread(s), 76.73%. 13.96s left...)
Average CPU speed: 77.47MH/s (8 thread(s), 80.07%. 11.96s left...)
Average CPU speed: 76.85MH/s (8 thread(s), 83.40%. 9.96s left...)
Average CPU speed: 76.98MH/s (8 thread(s), 86.73%. 7.96s left...)
Average CPU speed: 77.08MH/s (8 thread(s), 90.07%. 5.96s left...)
Average CPU speed: 77.18MH/s (8 thread(s), 93.41%. 3.96s left...)
Average CPU speed: 77.33MH/s (8 thread(s), 96.74%. 1.96s left...)
Average CPU speed: 77.42MH/s (8 thread(s), 100.00%. 0.00ns left...)
Average CPU speed: 77.48MH/s (8 thread(s), took 60.00s)
```

In comparision, this is what I get from the hornet node:

```
./hornet tools bench-cpu
Average CPU speed: 4.10MH/s (8 thread(s), 3.40%. 57s left...)
Average CPU speed: 4.03MH/s (8 thread(s), 6.75%. 55s left...)
Average CPU speed: 4.06MH/s (8 thread(s), 10.12%. 53s left...)
Average CPU speed: 4.07MH/s (8 thread(s), 13.49%. 51s left...)
Average CPU speed: 4.06MH/s (8 thread(s), 16.86%. 49s left...)
Average CPU speed: 4.05MH/s (8 thread(s), 20.23%. 47s left...)
Average CPU speed: 4.05MH/s (8 thread(s), 23.57%. 45s left...)
Average CPU speed: 4.05MH/s (8 thread(s), 26.92%. 43s left...)
Average CPU speed: 4.05MH/s (8 thread(s), 30.32%. 41s left...)
Average CPU speed: 4.05MH/s (8 thread(s), 33.68%. 39s left...)
Average CPU speed: 4.04MH/s (8 thread(s), 37.03%. 37s left...)
Average CPU speed: 4.04MH/s (8 thread(s), 40.38%. 35s left...)
Average CPU speed: 4.02MH/s (8 thread(s), 43.74%. 33s left...)
Average CPU speed: 4.00MH/s (8 thread(s), 47.12%. 31s left...)
Average CPU speed: 3.98MH/s (8 thread(s), 50.46%. 29s left...)
Average CPU speed: 3.98MH/s (8 thread(s), 53.80%. 27s left...)
Average CPU speed: 3.98MH/s (8 thread(s), 57.16%. 25s left...)
Average CPU speed: 3.98MH/s (8 thread(s), 60.51%. 23s left...)
Average CPU speed: 3.98MH/s (8 thread(s), 63.86%. 21s left...)
Average CPU speed: 3.97MH/s (8 thread(s), 67.23%. 19s left...)
Average CPU speed: 3.96MH/s (8 thread(s), 70.58%. 17s left...)
Average CPU speed: 3.96MH/s (8 thread(s), 73.95%. 15s left...)
Average CPU speed: 3.96MH/s (8 thread(s), 77.30%. 13s left...)
Average CPU speed: 3.96MH/s (8 thread(s), 80.67%. 11s left...)
Average CPU speed: 3.96MH/s (8 thread(s), 84.04%. 9s left...)
Average CPU speed: 3.95MH/s (8 thread(s), 87.40%. 7s left...)
Average CPU speed: 3.95MH/s (8 thread(s), 90.74%. 5s left...)
Average CPU speed: 3.95MH/s (8 thread(s), 94.11%. 3s left...)
Average CPU speed: 3.95MH/s (8 thread(s), 97.46%. 1s left...)
Average CPU speed: 3.95MH/s (8 thread(s), took 1m0s)
```

These scores almost look to good to be true. Is there something obvious jumping at you?

The improvements seem to come mostly from restructuring the benchmark and have less to do with the changes to `bee-ternary`. This is the restructured benchmark with an older version of bee ternary:

```
cargo run --release -- bench-cpu
Benchmarking CPU with 8 threads
Average CPU speed: 79.28MH/s (8 thread(s), 3.33%. 58.01s left...)
Average CPU speed: 77.80MH/s (8 thread(s), 6.67%. 56.01s left...)
Average CPU speed: 77.88MH/s (8 thread(s), 10.01%. 54.00s left...)
Average CPU speed: 77.83MH/s (8 thread(s), 13.35%. 52.00s left...)
Average CPU speed: 78.05MH/s (8 thread(s), 16.68%. 49.99s left...)
Average CPU speed: 78.10MH/s (8 thread(s), 20.02%. 47.99s left...)
Average CPU speed: 78.24MH/s (8 thread(s), 23.35%. 45.99s left...)
Average CPU speed: 78.28MH/s (8 thread(s), 26.68%. 43.99s left...)
Average CPU speed: 78.33MH/s (8 thread(s), 30.02%. 41.99s left...)
Average CPU speed: 78.44MH/s (8 thread(s), 33.36%. 39.99s left...)
Average CPU speed: 78.49MH/s (8 thread(s), 36.70%. 37.98s left...)
Average CPU speed: 78.58MH/s (8 thread(s), 40.03%. 35.98s left...)
Average CPU speed: 78.64MH/s (8 thread(s), 43.37%. 33.98s left...)
Average CPU speed: 78.69MH/s (8 thread(s), 46.71%. 31.98s left...)
Average CPU speed: 78.76MH/s (8 thread(s), 50.05%. 29.97s left...)
Average CPU speed: 78.78MH/s (8 thread(s), 53.38%. 27.97s left...)
Average CPU speed: 78.82MH/s (8 thread(s), 56.73%. 25.96s left...)
Average CPU speed: 78.84MH/s (8 thread(s), 60.07%. 23.96s left...)
Average CPU speed: 78.82MH/s (8 thread(s), 63.41%. 21.96s left...)
Average CPU speed: 78.85MH/s (8 thread(s), 66.75%. 19.95s left...)
Average CPU speed: 78.86MH/s (8 thread(s), 70.08%. 17.95s left...)
Average CPU speed: 78.89MH/s (8 thread(s), 73.42%. 15.95s left...)
Average CPU speed: 78.91MH/s (8 thread(s), 76.75%. 13.95s left...)
Average CPU speed: 78.95MH/s (8 thread(s), 80.08%. 11.95s left...)
Average CPU speed: 78.97MH/s (8 thread(s), 83.42%. 9.95s left...)
Average CPU speed: 78.99MH/s (8 thread(s), 86.75%. 7.95s left...)
Average CPU speed: 79.02MH/s (8 thread(s), 90.09%. 5.95s left...)
Average CPU speed: 78.93MH/s (8 thread(s), 93.42%. 3.95s left...)
Average CPU speed: 78.63MH/s (8 thread(s), 96.76%. 1.94s left...)
Average CPU speed: 78.34MH/s (8 thread(s), 100.00%. 0.00ns left...)
Average CPU speed: 78.42MH/s (8 thread(s), took 60.00s)
```

**Comment by grtlr** [comment]
Closing in favor of #843.

**Pull request closed without merging**